client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -132,17 +134,22 @@ private JwtTokenProviderImpl(Controller controllerClient, String scopeName, Stri
     @Override
     public CompletableFuture<String> retrieveToken() {
         DelegationToken currentToken = this.delegationToken.get();
-
+        final CompletableFuture<String> result;
         if (currentToken == null) {
-            return this.refreshToken();
+            result = this.refreshToken();
         } else if (currentToken.getExpiryTime() == null) {
-            return CompletableFuture.completedFuture(currentToken.getValue());
+            result = CompletableFuture.completedFuture(currentToken.getValue());
+        } else if (this.tokenExpirySignal.get()) {
+            log.debug("Token was signaled as expired for scope/stream {}/{}", this.scopeName, this.streamName);
+            result = refreshToken();
+            this.tokenExpirySignal.compareAndSet(true, false);
```

Comment:
This should be done inside of refreshToken in case on of the other branches occurs right before this is set.